File: matlab/src/lang/+State/toMarginal.m
function [ni, nir, sir, kir] = toMarginal(sn, ind, state_i, phasesz, phaseshift, space_buf, space_srv, space_var) %#ok<INUSD>
% TOMARGINAL Compute marginal distributions for a specific node
%
% [NI, NIR, SIR, KIR] = TOMARGINAL(SN, IND, STATE_I, PHASESZ, PHASESHIFT, SPACE_BUF, SPACE_SRV, SPACE_VAR)
%
% @brief Extracts marginal job statistics from global state information for a specific node
%
% This function processes the global network state to compute marginal 
% distributions and job counts for a specific node, considering different
% scheduling strategies, service phases, and node types.
%
% @param sn Network structure or Network object
% @param ind Node index to extract marginal information for
% @param state_i Global state matrix or vector
% @param phasesz Vector of phase sizes for each class
% @param phaseshift Phase shift parameters for state extraction
% @param space_buf Buffer space configuration
% @param space_srv Service space configuration  
% @param space_var Variable space configuration
%
% @return ni Total jobs in node IND
% @return nir Total jobs per class in node IND [matrix: states x classes]
% @return sir Total jobs in service per class in node IND [matrix: states x classes]
% @return kir Total jobs in service per class and per phase in node IND [3D array: states x classes x phases]

% Copyright (c) 2012-2026, Imperial College London
% All rights reserved.

if ~isstruct(sn) % the input can be a Network object too
    sn=sn.getStruct();
end

% Join stations on FJ-augmented structs: plain per-class count state,
% no buffer/phase split and no service (jobs wait for synchronization)
if isfield(sn,'isfjaugmented') && sn.isfjaugmented && sn.nodetype(ind) == NodeType.Join
    R = sn.nclasses;
    if nargin < 3
        state_i = sn.state{sn.nodeToStateful(ind)};
    end
    nir = state_i(:,(end-R+1):end);
    sir = zeros(size(nir));
    kir = zeros(size(state_i,1),R,1);
    ni = sum(nir,2);
    return
end

% ind: node index
if ~sn.isstation(ind) && sn.isstateful(ind) % if stateful node
    if sn.nodetype(ind) == NodeType.Transition
        R = sn.nodeparam{ind}.nmodes;
        if nargin < 5
            % Initialize phasesz/phaseshift for Transition nodes from firingphases
            phasesz = sn.nodeparam{ind}.firingphases;
            if any(isnan(phasesz))
                phasesz = zeros(1, R);
                for m = 1:R
                    if iscell(sn.nodeparam{ind}.firingproc) && ~isempty(sn.nodeparam{ind}.firingproc{m})
                        phasesz(m) = size(sn.nodeparam{ind}.firingproc{m}{1}, 1);
                    else
                        phasesz(m) = 1;
                    end
                end
            end
            phaseshift = [0,cumsum(phasesz(1:end-1))];
        end
        if nargin < 8
            nmodes = R;
            space_buf = state_i(:,1:nmodes);
            space_srv = state_i(:,(nmodes+1):(nmodes+sum(phasesz)));
            space_var = state_i(:,(nmodes+sum(phasesz)+1):end);
        end
        sir = zeros(size(state_i,1),R); % class-r jobs in service
        kir = zeros(size(state_i,1),R,max(phasesz)); % class-r jobs in service in phase k
        for r=1:R
            for k=1:phasesz(r)
                kir(:,r,k) = space_srv(:,phaseshift(r)+k);
                sir(:,r) = sir(:,r) + kir(:,r,k);
            end
        end
        nir = sir;
        ni = sum(nir);
    else
        ni = sum(state_i(1:(end-sum(sn.nvars(ind,:)))));
        nir = state_i(1:(end-sum(sn.nvars(ind,:))));
        sir = nir; % jobs in service
        kir = sir; % jobs per phase
    end
    return
end

R = sn.nclasses;

ist = sn.nodeToStation(ind);
isf = sn.nodeToStateful(ind);

if nargin < 3
    state_i = sn.state{isf};
end

if nargin < 5
    phasesz = sn.phasessz(ist,:);
    phaseshift = sn.phaseshift(ist,:);
end

if sn.sched(ist) == SchedStrategy.PAS
    % PAS/OI: the state is the ordered list of class indices (no server split).
    Vp = sum(sn.nvars(ind,:));
    listcols = state_i(:,1:(end-Vp));
    nir = zeros(size(state_i,1),R);
    for r=1:R
        nir(:,r) = sum(listcols==r,2);
    end
    % Jobs in service = positions receiving a nonzero rate increment
    % Delta_mu(c1..cp) > 0 (the rest wait); utilization derives from this.
    muFun = sn.nodeparam{ind}.svcRateFun;
    sir = zeros(size(state_i,1),R);
    if ~isempty(muFun)
        for row=1:size(listcols,1)
            c = listcols(row, listcols(row,:)>0);
            muPrev = 0;
            for p=1:numel(c)
                muCur = muFun(c(1:p));
                if muCur - muPrev > 0
                    sir(row,c(p)) = sir(row,c(p)) + 1;
                end
                muPrev = muCur;
            end
        end
    else
        sir = nir;
    end
    kir = zeros(size(state_i,1),R,max(phasesz));
    kir(:,:,1) = sir;
    ni = sum(nir,2);
    return
end

isExponential = false;
if max(phasesz)==1
    isExponential = true;
end

if nargin < 8
    space_var = state_i(:,(end-sum(sn.nvars(ind,:))+1):end); % local variables
    space_srv = state_i(:,(end-sum(phasesz)-sum(sn.nvars(ind,:))+1):(end-sum(sn.nvars(ind,:)))); % server state
    space_buf = state_i(:,1:(end-sum(phasesz)-sum(sn.nvars(ind,:)))); % buffer state
end

if isExponential
    sir = space_srv;
    kir = space_srv;
else
    nir = zeros(size(state_i,1),R);
    sir = zeros(size(state_i,1),R); % class-r jobs in service
    kir = zeros(size(state_i,1),R,max(phasesz)); % class-r jobs in service in phase k
    for r=1:R
        for k=1:phasesz(r)
            kir(:,r,k) = space_srv(:,phaseshift(r)+k);
            sir(:,r) = sir(:,r) + kir(:,r,k);
        end
    end
end

switch sn.sched(ist)
    case SchedStrategy.INF
        for r=1:R
            nir(:,r) = sir(:,r); % class-r jobs in station
        end
    case {SchedStrategy.PS, SchedStrategy.PSPRIO}
        for r=1:R
            nir(:,r) = sir(:,r) ; % class-r jobs in station
        end
    case SchedStrategy.EXT
        for r=1:R
            nir(:,r) = Inf;
        end
    case {SchedStrategy.FCFS, SchedStrategy.FCFSPRIO, SchedStrategy.HOL}
        for r=1:R
            nir(:,r) = sir(:,r) + sum(space_buf==r,2); % class-r jobs in station
        end
    case {SchedStrategy.DPS, SchedStrategy.DPSPRIO}
        for r=1:R
            nir(:,r) = sir(:,r) ; % class-r jobs in station
        end
    case {SchedStrategy.GPS, SchedStrategy.GPSPRIO}
        for r=1:R
            nir(:,r) = sir(:,r) ; % class-r jobs in station
        end
    case {SchedStrategy.LCFS, SchedStrategy.LCFSPRIO}
        for r=1:R
            nir(:,r) = sir(:,r) + sum(space_buf==r,2); % class-r jobs in station
        end
    case {SchedStrategy.FCFSPI, SchedStrategy.FCFSPIPRIO, SchedStrategy.FCFSPR, SchedStrategy.FCFSPRPRIO, SchedStrategy.LCFSPI, SchedStrategy.LCFSPIPRIO, SchedStrategy.LCFSPR, SchedStrategy.LCFSPRPRIO}
        % buffer holds (class,phase) pairs: slice the class columns per row.
        % length()+linear indexing would flatten a multi-row state matrix.
        if size(space_buf,2)>1
            space_bufcls = space_buf(:,1:2:end);
            %space_bufphase = space_buf(:,2:2:end);
            for r=1:R
                nir(:,r) = sir(:,r) + sum(space_bufcls==r,2); % class-r jobs in station
            end
        else
            nir = sir;
        end
    case {SchedStrategy.SIRO, SchedStrategy.POLLING}
        for r=1:R
            nir(:,r) = sir(:,r) + space_buf(:,r); % class-r jobs in station
        end
    case SchedStrategy.SEPT
        for r=1:R
            nir(:,r) = sir(:,r) + space_buf(:,r); % class-r jobs in station
        end
    case SchedStrategy.LEPT
        for r=1:R
            nir(:,r) = sir(:,r) + space_buf(:,r); % class-r jobs in station
        end
    case SchedStrategy.SRPT
        for r=1:R
            nir(:,r) = sir(:,r) + space_buf(:,r); % class-r jobs in station
        end
    case SchedStrategy.SRPTPRIO
        for r=1:R
            nir(:,r) = sir(:,r) + space_buf(:,r); % class-r jobs in station
        end
    otherwise % possibly other stateful nodes
        for r=1:R
            nir(:,r) = sir(:,r) ; % class-r jobs in station
        end
end

if sn.nodetype(ind) == NodeType.Place
    % A Place is a token container: the [buffer, server] split in its state
    % encoding is an artifact (a transition FIRE relocates the surviving
    % tokens from the server slot into the buffer slot). For the INF-family
    % scheduling used by Places the switch above set nir = sir (server slot
    % only), so fold the buffer slot back in. This makes the marginal token
    % count agree with State.toMarginalAggr (buf + srv), which drives the
    % transition enabling degree; without it the QLen measurement collapses
    % to the oscillating server slot while the dynamics stay correct.
    switch sn.sched(ist)
        case {SchedStrategy.INF, SchedStrategy.PS, SchedStrategy.PSPRIO, ...
              SchedStrategy.DPS, SchedStrategy.DPSPRIO, ...
              SchedStrategy.GPS, SchedStrategy.GPSPRIO, SchedStrategy.LPS}
            if ~isempty(space_buf) && size(space_buf,2) >= R
                for r=1:R
                    nir(:,r) = sir(:,r) + space_buf(:,r);
                end
            end
    end
    % set all service and phase data to 0
    for r=1:R
        if isnan(sn.rates(ist,r)) % if disabled station
            for k=1:phasesz(r)
                kir(:,r,k) = 0;
            end
            sir(:,r)=0;
        end
    end
else
    for r=1:R
        if isnan(sn.rates(ist,r)) % if disabled station
            nir(:,r) = 0;
            for k=1:phasesz(r)
                kir(:,r,k) = 0;
            end
            sir(:,r)=0;
        end
    end
end

ni = sum(nir,2); % total jobs in station
end

File: jar/src/main/kotlin/jline/lang/state/ToMarginal.java
package jline.lang.state;

import static jline.GlobalConstants.Inf;

import jline.lang.NetworkStruct;
import jline.lang.constant.NodeType;
import jline.lang.nodeparam.TransitionNodeParam;
import jline.util.matrix.Matrix;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;

import static jline.io.InputOutputKt.line_error;
import static jline.io.InputOutputKt.mfilename;

public class ToMarginal implements Serializable {
    /**
     * Computes marginal statistics for a stateful node in a Stochastic Petri Net.
     * 
     * This method extracts marginal statistics from the global state space, focusing on
     * a specific node and computing its local state information including job populations,
     * server allocations, and phase distributions. It handles different node types
     * (transitions, places, caches) and their specific state representations.
     * 
     * <p>For transition nodes with multi-phase service, it computes the distribution
     * of jobs across different service phases. For station nodes, it aggregates
     * queue lengths and server utilizations. The method is essential for event
     * processing in SPN models where local node state must be extracted from
     * the global state space.
     * 
     * @param sn Network structure containing the complete SPN model definition
     * @param ind Index of the node for which marginal statistics are computed
     * @param state_i Current state vector for this node from the global state space
     * @param phasesz Number of phases for each service mode (for transition nodes)
     * @param phaseshift Cumulative phase counts for indexing (for transition nodes)
     * @param space_buf Buffer space matrix containing queue state definitions
     * @param space_srv Server space matrix containing server allocation states
     * @param space_var Variable space matrix containing phase variable definitions
     * 
     * @return StateMarginalStatistics containing:
     *         <ul>
     *         <li>ni: Total number of jobs at the node</li>
     *         <li>nir: Number of jobs per class</li>
     *         <li>sir: Number of jobs in service per class</li>
     *         <li>kir: Per-phase job distribution for each class</li>
     *         </ul>
     * 
     * @see State.StateMarginalStatistics for the returned data structure
     * @see #toMarginalAggr for aggregated marginal computation variant
     */
    public static State.StateMarginalStatistics toMarginal(NetworkStruct sn, int ind, Matrix state_i, Matrix phasesz, Matrix phaseshift, Matrix space_buf, Matrix space_srv, Matrix space_var) {
        /* ------------------------------------------------------------------
         *  Cached (non-station) stateful node
         * ------------------------------------------------------------------ */
        if (sn.isstation.get(ind, 0) == 0 && sn.isstateful.get(ind, 0) > 0) {

            /* --- Transition node (multi-phase service) -------------------- */
            if (sn.nodetype.get(ind) == NodeType.Transition) {

                /* number of service modes for this transition */
                int R = ((TransitionNodeParam) sn.nodeparam.get(sn.nodes.get(ind))).nmodes;
                int numRows = state_i.getNumRows();

                // kir[mode] is a matrix of size numRows × numPhases for that mode
                // MATLAB: kir = zeros(size(state_i,1), R, max(phasesz))
                Matrix sir = new Matrix(numRows, R);
                List<Matrix> kir = new ArrayList<Matrix>();

                for (int r = 0; r < R; r++) {
                    int numPh = (int) phasesz.get(0, r);
                    Matrix kir_r = new Matrix(numRows, numPh);

                    for (int k = 0; k < numPh; k++) {
                        int col = (int) phaseshift.get(0, r) + k;
                        for (int row = 0; row < numRows; row++) {
                            double val = space_srv.get(row, col);
                            kir_r.set(row, k, val);
                            sir.set(row, r, sir.get(row, r) + val);
                        }
                    }
                    kir.add(kir_r);
                }

                Matrix nir = sir.copy();
                Matrix ni = nir.sumRows();

                return new State.StateMarginalStatistics(ni, nir, sir, kir);
            }

            /* --- Cache node ------------------------------------------------ */
            if (sn.nodetype.get(ind) == NodeType.Cache) {
                // For cache nodes, the state vector contains:
                // [job counts for each class, cache item states]
                // The job counts are the first sn.nclasses elements
                int R = sn.nclasses;
                Matrix nir = new Matrix(1, R);
                double niVal = 0.0;
                
                // Extract job counts per class (first R elements of state_i)
                for (int r = 0; r < R && r < state_i.length(); r++) {
                    double v = state_i.get(r);
                    nir.set(0, r, v);
                    niVal += v;
                }
                
                Matrix ni = new Matrix(1, 1);
                ni.set(0, 0, niVal);
                
                Matrix sir = nir.copy();  // for cache, all jobs are "in service"
                List<Matrix> kir = new ArrayList<Matrix>();
                kir.add(sir);
                
                return new State.StateMarginalStatistics(ni, nir, sir, kir);
            }

            /* --- Generic stateful (non-transition) node ------------------- */
            double nvarsSum = sn.nvars.sumRows(ind);
            int nCols = (int) (state_i.length() - nvarsSum);

            // Guard against negative column count
            if (nCols < 0) {
                line_error(mfilename(new Object[]{}), "Warning: nCols calculated as negative: " + nCols + 
                                 " (state_i.length=" + state_i.length() + ", nvarsSum=" + nvarsSum + ")");
                nCols = 0;
            }

            Matrix nir = new Matrix(1, Math.max(1, nCols));
            double niVal = 0.0;
            for (int i = 0; i < nCols && i < state_i.length(); i++) {
                double v = state_i.get(i);
                nir.set(0, i, v);
                niVal += v;
            }
            Matrix ni = new Matrix(1, 1);
            ni.set(0, 0, niVal);

            Matrix sir = nir;                // all jobs are “in service”
            List<Matrix> kir = new ArrayList<Matrix>();
            kir.add(sir);

            return new State.StateMarginalStatistics(ni, nir, sir, kir);
        }


        int R = sn.nclasses;
        int ist = (int) sn.nodeToStation.get(ind);

        if (phasesz == null) {
            phasesz = new Matrix(1, sn.phasessz.getNumCols());
            Matrix.extract(sn.phasessz, ist, ist + 1, 0, sn.phasessz.getNumCols(), phasesz, 0, 0);
        }
        if (phaseshift == null) {
            phaseshift = new Matrix(1, sn.phaseshift.getNumCols());
            Matrix.extract(sn.phaseshift, ist, ist + 1, 0, sn.phaseshift.getNumCols(), phaseshift, 0, 0);
        }

        boolean isExponential = (phasesz.elementMax() == 1);

        if (space_var == null) {
            int col = (int) sn.nvars.sumRows(ind);
            
            if (col == 0) {
                // No variables - create empty matrix
                space_var = new Matrix(state_i.getNumRows(), 0);
            } else {
                // Validate bounds to prevent matrix extraction errors
                int varStartCol = state_i.getNumCols() - col;
                if (varStartCol < 0 || col < 0 || varStartCol >= state_i.getNumCols()) {
                    // Handle edge case where variable space dimensions are invalid
                    space_var = new Matrix(state_i.getNumRows(), Math.max(0, col));
                    space_var.zero();
                } else {
                    space_var = new Matrix(state_i.getNumRows(), col);
                    Matrix.extract(state_i, 0, state_i.getNumRows(), varStartCol, state_i.getNumCols(), space_var, 0, 0);
                }
            }
        }
        if (space_srv == null) {
            int sumPhasesz = (int) phasesz.elementSum();
            int sumNvars = (int) sn.nvars.sumRows(ind);
            
            // Validate bounds to prevent matrix extraction errors
            int srvStartCol = state_i.getNumCols() - sumPhasesz - sumNvars;
            int srvEndCol = state_i.getNumCols() - sumNvars;
            
            if (srvStartCol < 0 || srvEndCol > state_i.getNumCols() || srvStartCol >= srvEndCol) {
                // Handle edge case where state_i dimensions don't match expected layout
                // This can happen with degenerate models or edge cases in state space generation
                space_srv = new Matrix(state_i.getNumRows(), Math.max(0, sumPhasesz));
                space_srv.zero();
            } else {
                space_srv = new Matrix(state_i.getNumRows(), sumPhasesz);
                Matrix.extract(state_i, 0, state_i.getNumRows(), srvStartCol, srvEndCol, space_srv, 0, 0);
            }
        }
        if (space_buf == null) {
            int col = state_i.getNumCols() - (int) (phasesz.elementSum() + sn.nvars.sumRows(ind));
            
            // Validate bounds to prevent matrix extraction errors
            if (col < 0 || col > state_i.getNumCols()) {
                // Handle edge case where calculated buffer size is invalid
                space_buf = new Matrix(state_i.getNumRows(), Math.max(0, col));
                space_buf.zero();
            } else {
                space_buf = new Matrix(state_i.getNumRows(), col);
                Matrix.extract(state_i, 0, state_i.getNumRows(), 0, col, space_buf, 0, 0);
            }
        }

        Matrix nir = new Matrix(state_i.getNumRows(), R);
        Matrix sir = new Matrix(state_i.getNumRows(), R);
        List<Matrix> kir = new ArrayList<Matrix>();
        if (isExponential) {
            sir = space_srv;
            kir.add(space_srv);
        } else {
            // Initialize kir
            for (int i = 0; i < phasesz.elementMax(); i++)
                kir.add(new Matrix(state_i.getNumRows(), R));

            for (int r = 0; r < R; r++) {
                for (int k = 0; k < phasesz.get(r); k++) {
                    // kir(:,r,k) = space_srv(:,phaseshift(r)+k);
                    Matrix tmp_kir = kir.get(k);
                    Matrix.extract(space_srv, 0, space_srv.getNumRows(), (int) phaseshift.get(r) + k, (int) phaseshift.get(r) + k + 1, tmp_kir, 0, r);

                    // sir(:,r) = sir(:,r) + kir(:,r,k);
                    for (int i = 0; i < sir.getNumRows(); i++)
                        sir.set(i, r, sir.get(i, r) + tmp_kir.get(i, r));
                }
            }
        }

        switch (sn.sched.get(sn.stations.get(ist))) {
            case INF:
            case PS:
            case DPS:
            case GPS:
            case LPS:
            case PSPRIO:
            case DPSPRIO:
            case GPSPRIO:
                nir = sir.copy();
                break;
            case EXT:
                nir.fill(Inf);
                break;
            case FCFS:
            case HOL:
            case FCFSPRIO:
            case LCFS:
            case LCFSPRIO:
                for (int r = 0; r < R; r++) {
                    Matrix sumval;
                    if (space_buf.getNumRows() == 0 || space_buf.getNumCols() == 0 || (space_buf.getNumRows() == 1 && space_buf.getNumCols() == 1 && space_buf.value() == 0)) {
                        sumval = new Matrix(1, 1);
                        sumval.set(0, 0, 0);
                    } else {
                        // +1 since loop starts from 0 but classes start from 1
                        sumval = space_buf.countEachRow(r + 1);
                    }
                    for (int i = 0; i < sir.getNumRows(); i++) nir.set(i, r, sir.get(i, r) + sumval.get(i));
                }
                break;
            case LCFSPI:
            case LCFSPR:
            case LCFSPIPRIO:
            case LCFSPRPRIO:
                if (space_buf.length() > 1) {
                    // space_buf = space_buf(1:2:end);
                    Matrix sub_space_buf = new Matrix(1, (space_buf.getNumCols() * space_buf.getNumRows() + 1) / 2);
                    for (int i = 0; i < sub_space_buf.getNumCols(); i++)
                        sub_space_buf.set(0, i, space_buf.get(i * 2));

                    for (int r = 0; r < R; r++) {
                        Matrix sumval = sub_space_buf.countEachRow(r);
                        for (int i = 0; i < sir.getNumRows(); i++) nir.set(i, r, sir.get(i, r) + sumval.get(i));
                    }
                } else {
                    nir = sir.copy();
                }
                break;
            case SIRO:
            case SEPT:
            case LEPT:
            case SRPT:
            case SRPTPRIO:
                for (int r = 0; r < R; r++) {
                    for (int i = 0; i < sir.getNumRows(); i++)
                        nir.set(i, r, sir.get(i, r) + space_buf.get(i, r));
                }
                break;
            case POLLING:
                // For polling servers, treat like FCFS with buffer counting
                for (int r = 0; r < R; r++) {
                    Matrix sumval;
                    if (space_buf.getNumRows() == 0 || space_buf.getNumCols() == 0 || (space_buf.getNumRows() == 1 && space_buf.getNumCols() == 1 && space_buf.value() == 0)) {
                        sumval = new Matrix(1, 1);
                        sumval.set(0, 0, 0);
                    } else {
                        // +1 since loop starts from 0 but classes start from 1
                        sumval = space_buf.countEachRow(r + 1);
                    }
                    for (int i = 0; i < sir.getNumRows(); i++) nir.set(i, r, sir.get(i, r) + sumval.get(i));
                }
                break;
            default:
                for (int r = 0; r < R; r++) {
                    for (int i = 0; i < sir.getNumRows(); i++) nir.set(i, r, sir.get(i, r));
                }
        }

        if (sn.nodetype.get(ind) != NodeType.Place) {
            for (int r = 0; r < R; r++) {
                if (Double.isNaN(sn.rates.get(ist, r))) {
                    for (int i = 0; i < nir.getNumRows(); i++) {
                        nir.remove(i, r);
                        sir.remove(i, r);
                    }
                    for (int k = 0; k < phasesz.get(r); k++) {
                        for (int j = 0; j < kir.size(); j++) {
                            kir.get(j).remove(r, k);
                        }
                    }
                }
            }
        }

        Matrix ni = nir.sumRows();

        return new State.StateMarginalStatistics(ni, nir, sir, kir);
    }

    /**
     * Computes aggregated marginal statistics for a stateful node using specified aggregation weights.
     * 
     * This method is a specialized variant of {@link #toMarginal} that applies aggregation weights
     * to compute weighted marginal statistics. It is particularly useful in event processing
     * where multiple state configurations need to be combined with specific probabilities or
     * weights, such as during transition enabling calculations or place token aggregation.
     * 
     * <p>The aggregation weights K and Ks determine how different components of the state
     * space are combined to produce the final marginal statistics. This is essential for
     * handling complex state dependencies in SPN models where the effective state at a
     * node depends on weighted contributions from multiple sources.
     * 
     * @param sn Network structure containing the complete SPN model definition
     * @param ind Index of the node for which aggregated marginal statistics are computed
     * @param state_i Current state vector for this node from the global state space
     * @param K Aggregation weight matrix for combining state components
     * @param Ks Cumulative aggregation weights for proper normalization
     * @param space_buf Buffer space matrix containing queue state definitions
     * @param space_srv Server space matrix containing server allocation states
     * @param space_var Variable space matrix containing phase variable definitions
     * 
     * @return StateMarginalStatistics containing weighted/aggregated:
     *         <ul>
     *         <li>ni: Total weighted number of jobs at the node</li>
     *         <li>nir: Weighted number of jobs per class</li>
     *         <li>sir: Weighted number of jobs in service per class</li>
     *         <li>kir: Weighted per-phase job distribution for each class</li>
     *         </ul>
     * 
     * @see State.StateMarginalStatistics for the returned data structure
     * @see #toMarginal for the non-aggregated variant
     * @see State#handleEnableEvent for usage in transition enabling calculations
     */
    public static State.StateMarginalStatistics toMarginalAggr(NetworkStruct sn, int ind, Matrix state_i, Matrix K, Matrix Ks, Matrix space_buf, Matrix space_srv, Matrix space_var) {

        int ist = (int) sn.nodeToStation.get(ind);
        int R = sn.nclasses;

        if (sn.isstation.get(ind, 0) == 0 && sn.isstateful.get(ind, 0) > 0) {
            // For non-station stateful nodes, return the sum of all state elements except the last nvars elements
            double nvarsSum = sn.nvars.sumRows(ind);
            int nCols = (int) (state_i.getNumCols() - nvarsSum);
            
            // Guard against negative column count
            if (nCols < 0) {
                nCols = 0;
            }
            
            Matrix nir = new Matrix(state_i.getNumRows(), Math.max(1, nCols));
            Matrix ni = new Matrix(state_i.getNumRows(), 1);
            
            for (int i = 0; i < state_i.getNumRows(); i++) {
                double niVal = 0.0;
                for (int j = 0; j < nCols && j < state_i.getNumCols(); j++) {
                    double v = state_i.get(i, j);
                    if (j < nir.getNumCols()) {
                        nir.set(i, j, v);
                    }
                    niVal += v;
                }
                ni.set(i, 0, niVal);
            }
            
            return new State.StateMarginalStatistics(ni, nir, null, null);
        }


        if (K == null) {
            K = new Matrix(1, sn.phasessz.getNumCols());
            Matrix.extract(sn.phasessz, ist, ist + 1, 0, sn.phasessz.getNumCols(), K, 0, 0);
        }
        if (Ks == null) {
            Ks = new Matrix(1, sn.phaseshift.getNumCols());
            Matrix.extract(sn.phaseshift, ist, ist + 1, 0, sn.phaseshift.getNumCols(), Ks, 0, 0);
        }

        if (space_var == null) {
            int col = (int) sn.nvars.sumRows(ind);
            if (col == 0) {
                // No variables - create empty matrix
                space_var = new Matrix(state_i.getNumRows(), 0);
            } else {
                space_var = new Matrix(state_i.getNumRows(), col);
                Matrix.extract(state_i, 0, state_i.getNumRows(), state_i.getNumCols() - col, state_i.getNumCols(), space_var, 0, 0);
            }
        }

        if (space_srv == null) {
            int sumK = (int) K.elementSum();
            int sumNvars = (int) sn.nvars.sumRows(ind);
            int startCol = state_i.getNumCols() - sumK;
            int endCol = state_i.getNumCols() - sumNvars;
            int numCols = endCol - startCol;
            space_srv = new Matrix(state_i.getNumRows(), numCols);
            Matrix.extract(state_i, 0, state_i.getNumRows(), startCol, endCol, space_srv, 0, 0);
        }

        if (space_buf == null) {
            int col = state_i.getNumCols() - (int) (K.elementSum());
            space_buf = new Matrix(state_i.getNumRows(), col);
            Matrix.extract(state_i, 0, state_i.getNumRows(), 0, col, space_buf, 0, 0);
        }
        Matrix nir = new Matrix(state_i.getNumRows(), R);
        nir.zero();
        for (int r = 0; r < R; r++) {
            for (int k = 0; k < K.get(r); k++) {
                Matrix tmp_kir = new Matrix(state_i.getNumRows(), 1);
                Matrix.extract(space_srv, 0, space_srv.getNumRows(), (int) Ks.get(r) + k, (int) Ks.get(r) + k + 1, tmp_kir, 0, 0);
                for (int i = 0; i < nir.getNumRows(); i++) {
                    nir.set(i, r, nir.get(i, r) + tmp_kir.get(i));
                }
            }
        }

        // MATLAB LINE does not handle LCFSPR, INF, PS, DPS, GPS cases
        // nir: class-r jobs at the station
        switch (sn.sched.get(sn.stations.get(ist))) {
            case EXT:
                nir.fill(Inf);
                break;
            case FCFS:
            case HOL:
            case FCFSPRIO:
            case LCFS:
            case LCFSPRIO:
                for (int r = 0; r < R; r++) {
                    Matrix sumval;
                    if (space_buf.getNumRows() == 0 || space_buf.getNumCols() == 0 || (space_buf.getNumRows() == 1 && space_buf.getNumCols() == 1 && space_buf.value() == 0)) {
                        sumval = new Matrix(1, 1);
                        sumval.set(0, 0, 0);
                    } else {
                        // +1 since loop starts from 0 but classes start from 1
                        sumval = space_buf.countEachRow(r + 1);
                    }
                    for (int i = 0; i < nir.getNumRows(); i++) nir.set(i, r, nir.get(i, r) + sumval.get(i));
                }
                break;
            case SIRO:
            case SEPT:
            case LEPT:
            case SRPT:
            case SRPTPRIO:
                for (int r = 0; r < R; r++) {
                    for (int i = 0; i < nir.getNumRows(); i++)
                        nir.set(i, r, nir.get(i, r) + space_buf.get(i, r));
                }
                break;
            default:
                // do nothing, no-op case
        }

        for (int r = 0; r < R; r++) {
            if (Double.isNaN(sn.rates.get(ist, r)) && sn.nodetype.get(ind) != NodeType.Place) { // if station disabled
                for (int i = 0; i < nir.getNumRows(); i++) {
                    nir.remove(i, r);
                }
            }
        }

        Matrix ni = nir.sumRows();
        return new State.StateMarginalStatistics(ni, nir, null, null);

    }
}